Data/models/DataController.swift

```diff
@@ -20,6 +20,37 @@ enum WriteContext {
 
 public class DataController: NSObject {
     private static let databaseName = "Brave.sqlite"
+    private static let modelName = "Model"
+    private let container: NSPersistentContainer
+    
+    // MARK: - Initialization
+    
+    /// Managed Object Model of the database stack.
+    /// Must be created only once, this is to prevent a bug when testing with in-memory store.
+    /// More info here https://stackoverflow.com/a/51857486.
+    /// Note: this might be not needed in Swift 5.1 or newer.
+    private static let model: NSManagedObjectModel = {
```

Comment:
See https://stackoverflow.com/questions/51851485/multiple-nsentitydescriptions-claim-nsmanagedobject-subclass/51857486#51857486 for more context

This is to prevent `warning: Multiple NSEntityDescriptions claim the NSManagedObject subclass 'Data.Bookmark' so +entity is unable to disambiguate` error in tests.

So this model is created only once, when previously it was created on every single test

Reply:
I got rid of this warning by using `-[NSEntityDescription entityForName:inManagedObjectContext:]` rather than `+entity` on the model itself.

Reply:
We use it too
```
static func entity(context: NSManagedObjectContext) -> NSEntityDescription {
        // Swift 4 version
        // let className = String(describing: type(of: self))
        let className = String(describing: self)
        return NSEntityDescription.entity(forEntityName: className, in: context)!
    }
```

same, right?

Reply:
Thats odd, yeah its the same. This got rid of all the warnings when running tests for rewards database

Reply:
Hmm but you shouldn't have more than one instance of PersistentContainer.
So the persistent container should be created only once and then make as many contexts as we want or add as many data-stores as we want.

You will get the same exact error for deleting a database or removing a persistent store if there is more than one container.

I think your original fix was correct in keeping it like a singleton.
I had the same problem for Google Safe Browsing and the above fixes in the comment didn't fix it for me. Only fix was making the persistent container a `let` constant and using it wherever needed.

If the DB got corrupted, I could keep the same container, and just redo the persistent store. But whenever I remade the container, I got write errors and corruption :(

Reply:
The problem should not be the ManagedObjectModel. Tests need to create the container only **once**. The managed object model can be created as many times as we want..

However, each test needs to tear down the persistent store loaded into the container if they want afaik.

As far as I'm aware, the problem is multiple containers and not the model.

I faced the exact same problem and I solved it with: https://github.com/brave/brave-ios/blob/5fa1fc9507128e83a5e8272267cf3a797a253805/Client/SafeBrowsing/GSafeBrowsingDatabase.swift#L296

With a single persistent container, I no longer had the problem.. but I still needed to reset the db to its original state and tear down the stack. To do that, I used iOS 9+ function `destroyPersistentStore`:

https://github.com/brave/brave-ios/blob/5fa1fc9507128e83a5e8272267cf3a797a253805/Client/SafeBrowsing/GSafeBrowsingDatabase.swift#L323

Reply:
What do you think about destroying the container in `deinit` method of `DataController`?

I don't like having one container for all tests that much because `NSPersistentContainer` manages `NSManagedObjectContexts` which may or may not be fully reseted, some state might be still present there etc. especially with that many background writes that we have in tests. 

I think it's better to just recreate whole DB stack in tests, and have fresh container with fresh objectcontexts, but doesn't hurt to add that `destroyPersistent` store in `deinit` method

This only applies to test environment and using in memory store of course.

Lmk what you think about it